### PR TITLE
feat: 0.14.2 bump + 0.10.2 RPC

### PR DESCRIPTION
**Pull request by Mohiiit**
## Summary
- add direct Madara JSON-RPC `0.10.2` support and register it as the latest user RPC version
- add the `0.10.2` RPC primitive/schema layer and REST method updates for storage, state update, events, messages status, and trace plumbing
- align transaction hashing and add sequencer/pathfinder-backed regression coverage for `0.10.2`

## Validation
- cargo fmt --all
- cargo clippy --workspace --all-targets --locked -- -D warnings
- cargo test -p mp-transactions compute_hash --lib
- cargo test -p mp-transactions compute_hash::tests::test_compute_hash_matches_sequencer_vectors --lib

### CODE REVIEW

File: orchestrator/src/worker/utils/fact_topology.rs [hunk 1]
```diff
@@ -65,7 +65,7 @@ pub fn get_fact_topology(cairo_pie: &CairoPie, output_size: usize) -> Result<Fac
 
 /// Returns the sizes of the program output pages, given the pages dictionary that appears
 /// in the additional attributes of the output builtin.
-pub fn get_page_sizes(pages: &HashMap<usize, PublicMemoryPage>, output_size: usize) -> Result<Vec<usize>, FactError> {
+pub fn get_page_sizes(pages: &BTreeMap<usize, PublicMemoryPage>, output_size: usize) -> Result<Vec<usize>, FactError> {
```
Comment (prkpndy) [inline orchestrator/src/worker/utils/fact_topology.rs:68]: why did we change this?

File: madara/crates/primitives/transactions/src/lib.rs [hunk 2]
```diff
@@ -1106,6 +1108,7 @@ mod tests {
             account_deployment_data: vec![Felt::from(10), Felt::from(11)],
             nonce_data_availability_mode: DataAvailabilityMode::L1,
             fee_data_availability_mode: DataAvailabilityMode::L2,
+            proof_facts: None,
```
Comment (prkpndy) [inline madara/crates/primitives/transactions/src/lib.rs:1122]: should we have some value here? assuming this is being used for testing purposes!

File: madara/crates/primitives/rpc/src/lib.rs [hunk 3]
```diff
@@ -2,6 +2,7 @@ mod custom_serde;
 
 pub mod admin;
 pub mod v0_10_0;
+pub mod v0_10_2;
```
Comment (prkpndy) [inline madara/crates/primitives/rpc/src/lib.rs:5]: btw, don't we have v0.10.1?

**Review by heemankv — COMMENTED** [review]
(review withdrawn)

File: madara/crates/client/block_production/src/executor/thread.rs [hunk 4]
```diff
@@ -136,23 +135,45 @@ impl ExecutorThread {
...
-                }
-                _ = OptionFuture::from(deadline.map(tokio::time::sleep_until)) => {
-                    tracing::debug!("Waiting for batch timed out.");
-                    WaitTxBatchOutcome::Batch(Default::default())
-                }
-                el = self.incoming_batches.recv() => match el {
-                    Some(el) => {
-                        tracing::debug!("Got new batch with {} transactions.", el.len());
-                        WaitTxBatchOutcome::Batch(el)
+            match deadline {
```
Comment (prkpndy) [inline madara/crates/client/block_production/src/executor/thread.rs:139]: Let's mention in the function doc that passing `None` => no timeout, wait indefinitely. was a little confusing for me in the first read!

File: madara/crates/client/block_production/src/lib.rs [hunk 5]
```diff
@@ -761,12 +761,31 @@ impl BlockProductionTask {
 
                 // Check if pre-confirmed block exists (it shouldn't at this point)
                 // Create new preconfirmed block
+                let preconfirmed_header = exec_ctx.into_header();
+                tracing::info!(
```
Comment (prkpndy) [inline madara/crates/client/block_production/src/lib.rs]: should we make this debug?

File: madara/crates/client/block_production/src/lib.rs [hunk 6]
```diff
@@ -761,12 +761,31 @@ impl BlockProductionTask {
...
+
                 let backend = self.backend.clone();
                 global_spawn_rayon_task(move || {
-                    backend.write_access().new_preconfirmed(PreconfirmedBlock::new(exec_ctx.into_header()))
+                    backend.write_access().new_preconfirmed(PreconfirmedBlock::new(preconfirmed_header))
                 })
                 .await?;
 
+                if let Some(preconfirmed) = self.backend.chain_tip.borrow().as_preconfirmed() {
+                    tracing::info!(
```
Comment (prkpndy) [inline madara/crates/client/block_production/src/lib.rs]: same

File: madara/crates/client/db/src/rocksdb/blocks.rs [hunk 7]
```diff
@@ -27,6 +32,114 @@ struct StateDiffV8 {
...
+    nonce: Felt,
+    resource_bounds: ResourceBoundsMapping,
+    tip: u64,
+    paymaster_data: Vec<Felt>,
+    account_deployment_data: Vec<Felt>,
+    nonce_data_availability_mode: DataAvailabilityMode,
+    fee_data_availability_mode: DataAvailabilityMode,
+}
+
+impl From<LegacyTransactionWithReceipt> for TransactionWithReceipt {
```
Comment (prkpndy) [inline madara/crates/client/db/src/rocksdb/blocks.rs:83]: Let's add tests for all the From/Into impl from Legacy -> Current, and vice-versa.
Also, let's test the normal/current-schema path of deserialize_transaction_with_receipt and where a current invoke-v3 row with proof_facts: Some(...) is decoded.

File: madara/crates/client/db/src/test_utils.rs [hunk 8]
```diff
@@ -68,6 +68,7 @@ pub fn invoke_v3(sender: Felt, nonce: Felt) -> InvokeTransactionV3 {
         account_deployment_data: vec![Felt::from_hex_unchecked("0xbeef")],
         nonce_data_availability_mode: DataAvailabilityMode::L1,
         fee_data_availability_mode: DataAvailabilityMode::L1,
+        proof_facts: None,
```
Comment (prkpndy) [inline madara/crates/client/db/src/test_utils.rs:71]: should we add proof_facts here to test with that?

File: madara/crates/client/db/src/view/mod.rs [hunk 9]
```diff
@@ -62,37 +62,52 @@ impl<D: MadaraStorageRead> MadaraBackend<D> {
...
-                            &quote,
-                            parent_block_info.header.gas_prices.strk_l2_gas_price,
-                            parent_block_info.total_l2_gas_used,
-                        )?
+                let (block_timestamp, gas_prices) =
+                    if let Some(custom_header) = self.get_custom_header(parent_block_number + 1) {
+                        // Convert Unix timestamp (seconds since Jan 1, 1970) to SystemTime
+                        let block_timestamp = UNIX_EPOCH + Duration::from_secs(custom_header.timestamp);
+                        let gas_prices = custom_header.gas_prices;
+                        tracing::info!(
```
Comment (prkpndy) [inline madara/crates/client/db/src/view/mod.rs]: should we make this debug?

File: madara/crates/client/mempool/src/inner/tests/mod.rs [hunk 10]
```diff
@@ -64,6 +64,7 @@ impl From<TestTx> for ValidatedTransaction {
                         account_deployment_data: Default::default(),
                         nonce_data_availability_mode: Default::default(),
                         fee_data_availability_mode: Default::default(),
+                        proof_facts: None,
```
Comment (prkpndy) [inline madara/crates/client/mempool/src/inner/tests/mod.rs]: nit: let's use default here to be consistent

File: madara/crates/client/rpc/src/types.rs [hunk 11]
```diff
@@ -74,4 +110,32 @@ mod tests {
...
+            event_index_in_block: 0,
+            in_preconfirmed: false,
+        }
+    }
+
+    #[rstest]
+    #[case(&[0, 0, 0], 2, ContinuationToken { block_number: 0, event_n: 0 }, Some(ContinuationToken { block_number: 0, event_n: 2 }))]
+    #[case(&[0, 0, 4], 2, ContinuationToken { block_number: 0, event_n: 0 }, Some(ContinuationToken { block_number: 4, event_n: 0 }))]
+    #[case(&[4, 6, 6], 2, ContinuationToken { block_number: 0, event_n: 2 }, Some(ContinuationToken { block_number: 6, event_n: 1 }))]
+    fn continuation_token_from_page_tracks_block_boundaries(
```
Comment (prkpndy) [inline madara/crates/client/rpc/src/types.rs:134]: continuation_token_from_page has useful coverage for the main pagination cases, but it is missing a few edge cases. In particular, page_size = 0 currently appears to produce a non-advancing continuation token (previous_token.event_n + 0) while build_events_chunk returns an empty page, which can make clients loop forever. Can we either reject chunk_size = 0 during request validation, or add an explicit test proving zero-sized pages make forward progress? I’d also add a simple direct test for events_infos.len() <= page_size returning None, since that is an important branch of this helper.

File: madara/crates/client/rpc/src/versions/user/common.rs [hunk 12]
```diff
@@ -0,0 +1,66 @@
+use crate::StarknetRpcApiError;
+use mp_convert::Felt;
+
+pub(crate) fn convert_storage_keys_for_v0_8_1(
```
Comment (prkpndy) [inline madara/crates/client/rpc/src/versions/user/common.rs:18]: This helper fixes the previous `unwrap_or(Felt::ZERO)` behavior, which is good, but the tests only cover a single valid key and a single invalid key. Can we add table coverage for `None`, empty input, multiple contract entries, multiple storage keys with order preserved, and mixed valid/invalid keys? I’d also like one `getStorageProof`-level test for an invalid storage key so we lock the RPC error shape. Also, since an invalid storage key is client input, returning `ErrUnexpectedError`/code 63 feels misleading; can we return an invalid-params style error or add a dedicated validation error instead?

File: madara/crates/primitives/transactions/src/to_blockifier.rs [hunk 13]
```diff
@@ -156,6 +156,17 @@ impl IntoStarknetApiExt for mp_rpc::v0_8_1::BroadcastedTxn {
     }
 }
 
+impl IntoStarknetApiExt for mp_rpc::v0_10_2::BroadcastedTxn {
+    fn into_starknet_api(
+        self,
+        chain_id: Felt,
+        starknet_version: StarknetVersion,
+    ) -> Result<(ApiAccountTransaction, Option<ConvertedClass>), ToBlockifierError> {
+        let tx: mp_rpc::v0_8_1::BroadcastedTxn = self.into();
```
Comment (prkpndy) [inline madara/crates/primitives/transactions/src/to_blockifier.rs]: This downcast drops v0.10.2 invoke-v3 `proof` / `proof_facts` before blockifier conversion. `simulateTransactions` and `estimateFee` use this path, so a v0.10.2 client can get results for a different transaction hash/context than the one it submitted.

File: madara/crates/client/rpc/src/versions/admin/v0_1_0/methods/write.rs [hunk 14]
```diff
@@ -272,7 +272,8 @@ impl MadaraWriteRpcApiV0_1_0Server for Starknet {
             .into());
         }
 
-        self.backend.set_custom_header(custom_block_headers);
+        self.backend.set_custom_header(custom_block_headers).map_err(StarknetRpcApiError::from)?;
```
Comment (prkpndy) [inline madara/crates/client/rpc/src/versions/admin/v0_1_0/methods/write.rs:275]: This returns success after staging the custom header, but if that block is already live as preconfirmed the live header and executor context are not updated; block close later only consumes the staged header for expected-hash comparison.

File: madara/crates/client/rpc/src/versions/user/v0_10_2/methods/read/get_events.rs [hunk 15]
```diff
@@ -0,0 +1,307 @@
...
+    )?;
+
+    Ok(build_events_chunk(events_infos, chunk_size, &requested_continuation_token))
+}
+
+fn validate_filter_limits(chunk_size: usize, keys: Option<&[Vec<Felt>]>) -> StarknetRpcResult<usize> {
+    if keys.map(|patterns| patterns.iter().map(|pattern| pattern.len()).sum::<usize>()).unwrap_or(0) > MAX_EVENTS_KEYS {
+        return Err(StarknetRpcApiError::TooManyKeysInFilter);
+    }
+    if chunk_size > MAX_EVENTS_CHUNK_SIZE {
```
Comment (prkpndy) [inline madara/crates/client/rpc/src/versions/user/v0_10_2/methods/read/get_events.rs]: `chunk_size = 0` is accepted here. With zero, the collector can fetch one event, `build_events_chunk` truncates the response to an empty page, and `continuation_token_from_page` can return `previous_token.event_n + 0`. That can produce a non-advancing continuation token.

File: madara/crates/client/rpc/src/versions/user/v0_10_2/methods/ws/starknet_unsubscribe.rs [hunk 16]
```diff
@@ -0,0 +1,9 @@
+use crate::errors::StarknetRpcResult;
+use crate::Starknet;
+
+/// Unsubscribes from a WebSocket subscription by ID.
+pub async fn starknet_unsubscribe(_starknet: &Starknet, _subscription_id: u64) -> StarknetRpcResult<bool> {
```
Comment (prkpndy) [inline madara/crates/client/rpc/src/versions/user/v0_10_2/methods/ws/starknet_unsubscribe.rs:5]: This implementation ignores the subscription id and always returns `Ok(true)`, so invalid ids are reported as successful and active subscriptions are not closed.

**Review by prkpndy — COMMENTED** [review]
some comments/questions. most importantly, i would suggest that we update the rpc test suite for testing the 14.2 implementation

File: orchestrator/src/worker/utils/fact_topology.rs (hunk 1 above)
Comment (Mohiiit) [reply orchestrator/src/worker/utils/fact_topology.rs:68]: was a upstream change in the cairoPIE itself, couldn't be helped!

File: madara/crates/primitives/transactions/src/lib.rs (hunk 2 above)
Comment (Mohiiit) [reply madara/crates/primitives/transactions/src/lib.rs:1122]: presenting right value would be hard, because it's a proof, will look for some txns if present on the sequencer repo!

File: madara/crates/primitives/rpc/src/lib.rs (hunk 3 above)
Comment (Mohiiit) [reply madara/crates/primitives/rpc/src/lib.rs:5]: 0.10.1 was introduced but within the day 0.10.2 was released, so 0.10.2 is 0.10.1 + additional changes (which will be used and no one will use 0.10.1)

File: madara/crates/client/rpc/src/versions/user/v0_10_2/methods/ws/starknet_unsubscribe.rs (hunk 16 above)
Comment (Mohiiit) [reply madara/crates/client/rpc/src/versions/user/v0_10_2/methods/ws/starknet_unsubscribe.rs:5]: ws is not implemented yet, so doesn't matter yet!

File: madara/crates/client/rpc/src/versions/admin/v0_1_0/methods/write.rs (hunk 14 above)
Comment (Mohiiit) [reply madara/crates/client/rpc/src/versions/admin/v0_1_0/methods/write.rs:275]: that's on the txn replay service to ensure the header is always sent first, and for the existing preconfirmed (shouldn't be the case because executor gets created only after first batch is passed)

for replay, this is fine IMO

File: madara/crates/client/block_production/src/util.rs [hunk 17]
```diff
@@ -191,12 +192,17 @@ pub(crate) fn create_execution_context(
...
     previous_l2_gas_used: u128,
 ) -> anyhow::Result<BlockExecutionContext> {
-    let (block_timestamp, gas_prices) = if let Some(custom_header) =
-        backend.get_custom_header().filter(|h| h.block_n == block_n)
-    {
+    let (block_timestamp, gas_prices) = if let Some(custom_header) = backend.get_custom_header(block_n) {
         // Convert Unix timestamp (seconds since Jan 1, 1970) to SystemTime
         let block_timestamp = UNIX_EPOCH + Duration::from_secs(custom_header.timestamp);
         let gas_prices = custom_header.gas_prices;
+        tracing::info!(
```
Comment (heemankv) [inline madara/crates/client/block_production/src/util.rs]: Do we change this to debug as well ? 
(following up on @prkpndy's request to turn custom header logs to debug)

I am fine with either debug or info, just looking for consistency

File: madara/crates/client/rpc/src/versions/user/v0_10_2/methods/read/get_storage_proof.rs [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+use mp_rpc::v0_10_0::BlockId;
+use mp_rpc::v0_10_2::{ContractStorageKeysItem, GetStorageProofResult};
+
+pub fn get_storage_proof(
+    starknet: &Starknet,
+    block_id: BlockId,
+    class_hashes: Option<Vec<Felt>>,
+    contract_addresses: Option<Vec<Felt>>,
+    contracts_storage_keys: Option<Vec<ContractStorageKeysItem>>,
+) -> StarknetRpcResult<GetStorageProofResult> {
```
Comment (heemankv) [inline madara/crates/client/rpc/src/versions/user/v0_10_2/methods/read/get_storage_proof.rs]: Madara will return the storage proofs at the latest confirmed block when we enter preconfirmed as the block tag, whereas it should give "Invalid block id"

**Review by heemankv — COMMENTED** [review]
Overall looks good, had some nitpicks but I've chosen to defer them for now, one comment added.

File: madara/crates/client/rpc/src/versions/user/v0_10_2/methods/ws/starknet_unsubscribe.rs (hunk 16 above)
Comment (prkpndy) [reply madara/crates/client/rpc/src/versions/user/v0_10_2/methods/ws/starknet_unsubscribe.rs:5]: Got it. Let's mention this in a comment then?

File: madara/crates/primitives/rpc/src/lib.rs (hunk 3 above)
Comment (prkpndy) [reply madara/crates/primitives/rpc/src/lib.rs:5]: got it

File: madara/crates/client/sync/src/import.rs [hunk 19]
```diff
@@ -400,25 +400,33 @@ impl BlockImporterCtx {
...
+                        return Err(BlockImportError::CompiledClassHash {
+                            class_hash,
+                            got: gateway_hash,
+                            expected: hashes.blake_hash,
+                        });
+                    }
+                    (None, Some(gateway_hash))
+                } else {
+                    if !self.config.no_check && hashes.poseidon_hash != gateway_hash {
+                        tracing::debug!(
```
Comment (heemankv) [inline madara/crates/client/sync/src/import.rs]: **Nitpick:** This should be `warn!` rather than `debug!` — a compiled class hash mismatch is worth operator visibility even if you're intentionally accepting the gateway value.

**Review by heemankv — COMMENTED** [review]
Good job, review and re-review done.

File: .prettierignore [hunk 20]
```diff
@@ -14,3 +14,4 @@ orchestrator/build/
 orchestrator/node_modules/
 lib/
 s3/
+bootstrapper-v2/contracts/ethereum/src/starkware/
```
Comment (prkpndy) [inline .prettierignore:17]: why do we need to add this?

File: orchestrator/src/utils/mod.rs [hunk 21]
```diff
@@ -8,7 +8,3 @@ pub mod metrics_recorder;
 pub mod preflight;
 pub mod rest_client;
 pub mod signal_handler;
-
-/// TODO: This is super Awkward to have this code here
-/// but will try to remove this and move it to the config from the root path
-pub const COMPILED_OS: &[u8] = include_bytes!("../../../build-artifacts/cairo_lang/os_latest.json");
```
Comment (prkpndy) [inline orchestrator/src/utils/mod.rs:14]: This was not being used anywhere, right? We can also update the build artifacts logic to not build/include this in the artifacts zip. That is redundant now.
If we are not doing that in this PR, let's add a TODO for that to pick later?

**Review by prkpndy — APPROVED** [review]
LGTM. No blockers.
Regarding chain config, I would suggest to update the version of that since the actual yaml has changed. We can later update the code itself to own the types used in the `ChainConfig` struct so that we can detect any changes at compile time only.
Also, I would suggest to make greptile score 5/5 before merging (or mention why the issues raised by it are not relevant)

File: orchestrator/src/utils/mod.rs (hunk 21 above)
Comment (Mohiiit) [reply orchestrator/src/utils/mod.rs:14]: sure!
